rest_server/RestServer.go

```diff
@@ -0,0 +1,269 @@
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+type RestServer struct {
+	Ipv4HTTPPort          uint64
+	Ipv4HTTPSPort         uint64
+	Ipv6HTTPPort          uint64
+	Ipv6HTTPSPort         uint64
+	TLSCertificate        string
+	TLSPrivateKey         string
+	HTTPChunkSize         uint64
+	logFileDestination    string
+	logFileSeverity       string
+	logConsoleDestination string
+	logConsoleSeverity    string
+}
+
+type Config map[string]string
+
+var (
+	restServer      RestServer
+	IOMgrCPool      *IOMgrConnPool
+	siodbLoggerPool SiodbLoggerPool
+)
+
+func main() {
+
+	SiodbInstanceName := flag.String("instance", "siodb", "Instance name")
+	flag.Parse()
+	SiodbInstanceConfigurationPath := "/etc/siodb/instances/" + *SiodbInstanceName
+	var err error
+	var config Config
+
+	// Parse Siodb instance parameters file
+	if config, err = loadInitFile(SiodbInstanceConfigurationPath + "/config"); err != nil {
+		log.Fatalln(err)
+	}
+
+	// Validate and load parameters
+	if err = restServer.loadConfig(SiodbInstanceConfigurationPath, config); err != nil {
+		log.Fatalln(err)
+	}
+
+	// Logging setup
+	siodbLoggerPool.AddSiodbLogger(restServer.logConsoleSeverity, restServer.logConsoleDestination)
+	siodbLoggerPool.AddSiodbLogger(restServer.logFileSeverity, restServer.logFileDestination)
+
+	siodbLoggerPool.Output(INFO, "Starting with config: %v", restServer)
+
+	// Gin Log at global level
+	siodbLoggerPool.ConfigGinLogger()
+
+	// Connection pool to IOMgr
+	IOMgrCPool, err = CreateIOMgrConnPool(4, 8, config)
+	if err != nil {
+		siodbLoggerPool.Output(FATAL, "Cannot create connection pool: %v", err)
+	}
+
+	// Start routers
+	if restServer.Ipv4HTTPPort != 0 {
+		go func() {
+			var restWorker RestWorker
+			if err := restWorker.router.New(restServer.Ipv4HTTPPort); err != nil {
+				siodbLoggerPool.Output(FATAL, "Cannot create router: %v", err)
+			}
+			if err := restWorker.CreateRouter(); err != nil {
+				siodbLoggerPool.Output(FATAL, "Cannot create route: %v", err)
+			}
+			if err := restWorker.StartHTTPRouter(); err != nil {
+				siodbLoggerPool.Output(FATAL, "Cannot start router: %v", err)
+			}
+		}()
+	}
+	if restServer.Ipv4HTTPSPort != 0 {
+		go func() {
+			var restWorker RestWorker
+			if err := restWorker.router.New(restServer.Ipv4HTTPSPort); err != nil {
+				siodbLoggerPool.Output(FATAL, "Cannot create router: %v", err)
+			}
+			if err := restWorker.CreateRouter(); err != nil {
+				siodbLoggerPool.Output(FATAL, "Cannot create route: %v", err)
+			}
+			if err := restWorker.StartHTTPSRouter(restServer.TLSCertificate, restServer.TLSPrivateKey); err != nil {
+				siodbLoggerPool.Output(FATAL, "Cannot start router: %v", err)
+			}
+		}()
+	}
+	if restServer.Ipv6HTTPPort != 0 {
+		go func() {
+			var restWorker RestWorker
+			if err := restWorker.router.New(restServer.Ipv6HTTPPort); err != nil {
```

Comment:
this should distinguish that this is IPv6. I suggest to add one more string parameter to "router" called "protocol", and for IPv4 you pass "tcp", and for IPv6 you pass "tcp6" and then use that for `Dial` / `Listen` calls. However, in this case you have to tell to "gin" that you want exactly "tcp6", it will not happen somehow automatically.

Reply:
@cxxman actually, the router listen use the underlaying `http.ListenAndServe`:
Which listen on both, ipv6 and ipv4: https://stackoverflow.com/questions/49929114/golang-ipv6-server
There is no way to indicate a protocol. What we can do later if we want to add to a specific address is to enclose the IPv6 in square brackets, as in "[2001:db8::1]:80" or "[fe80::1%zone]:80"as described here: https://golang.org/pkg/net/#Dial